core/frontend/src/render/webgl/Texture.ts

```diff
@@ -480,6 +511,88 @@ export class Texture2DHandle extends TextureHandle {
   }
 }
 
+/** @internal */
+export interface ExternalTextureRequest {
+  handle: Texture2DHandle;
+  name: string;
+  imodel: IModelConnection;
+  type: RenderTexture.Type;
+  format: ImageSourceFormat;
+  onLoaded?: (req: ExternalTextureRequest) => void;
+}
+
+/** @internal */
+export class ExternalTextureLoader { /* currently exported for tests only */
+  public static readonly instance = new ExternalTextureLoader(10);
+  private readonly _maxActiveRequests: number;
+  private _activeRequests: Array<ExternalTextureRequest> = [];
+  private _pendingRequests: Array<ExternalTextureRequest> = [];
+
+  public get numActiveRequests() { return this._activeRequests.length; }
+  public get numPendingRequests() { return this._pendingRequests.length; }
+  public get maxActiveRequests() { return this._maxActiveRequests; }
+
+  private constructor(maxActiveRequests: number) {
+    this._maxActiveRequests = maxActiveRequests;
+  }
+
+  private _nextRequest(prevReq: ExternalTextureRequest) {
+    this._activeRequests.splice(this._activeRequests.indexOf(prevReq), 1);
+    if (this._activeRequests.length < this._maxActiveRequests && this._pendingRequests.length > 0) {
+      const req = this._pendingRequests.pop()!;
+      this._activateRequest(req);
+    }
+  }
+
+  private _activateRequest(req: ExternalTextureRequest) {
+    this._activeRequests.push(req);
+    try {
+      const texBytesPromise = req.imodel.getTextureImage({ name: req.name });
+      texBytesPromise.then((texBytes: Uint8Array | undefined) => {
+        if (undefined !== texBytes && !req.imodel.isClosed) {
+          const imageSource = new ImageSource(texBytes, req.format);
+          const imageElementPromise = imageElementFromImageSource(imageSource);
+          imageElementPromise.then((image: HTMLImageElement) => {
+            if (!req.imodel.isClosed) {
+              req.handle.reload(Texture2DCreateParams.createForImage(image, ImageSourceFormat.Png === req.format, req.type));
+              IModelApp.tileAdmin.invalidateAllScenes();
+              if (undefined !== req.onLoaded)
+                req.onLoaded(req);
+            }
+            this._nextRequest(req);
+          }).catch((_e) => { });
+        } else
+          this._nextRequest(req);
+      }).catch((_e) => { });
+    } catch (_e) {
+      this._nextRequest(req); // if exception occurs while calling getTextureImage, remove this request and go to the next.
+    }
+  }
+
+  private _requestExists(reqToCheck: ExternalTextureRequest) {
+    for (const r of this._activeRequests)
+      if (reqToCheck.name === r.name && reqToCheck.imodel.iModelId === r.imodel.iModelId)

```

Comment:
`IModel.iModelId` is allowed to be undefined - is this comparison reliable.

Reply:
I was trying to check if the requests were for the same iModel -  I guess we could just compare the objects...

Reply:
Changed this to make it compare iModel objects.